show color scheme in skin preview

ronso0 opened the pull request:
I think I'm on the right track here. Though I need some help in [DlgPrefInterface::DlgPrefInterface](https://github.com/mixxxdj/mixxx/pull/2008/files#diff-23eb4e5036d4330d5058f3ab464e6579R96)

- [x] rename **preferences_skin_preview.png** > **skin_preview.png**
- [x] add **skin_preview_SchemeName.png** to Shade and Tango
- [x] move color scheme combobox directly below skin combobox

![scheme selector](https://user-images.githubusercontent.com/5934199/51788620-5f4e2c80-2180-11e9-99a9-de0fc5518e3e.png)

ronso0 committed:
move color scheme selector directly below skin selector

ronso0 committed:
make skin preview consider color scheme

ronso0 committed:
help comment

ronso0 commented:
I'll keep on trying, but I'd be glad if someone can suggest a solution quicker than me

ronso0 committed:
move updateSchemes into skin scanner loop

ronso0 commented on src/skin/skinloader.cpp:
```diff
@@ -56,12 +56,21 @@ QString SkinLoader::getSkinPath(const QString& skinName) const {
+    	QString schemeNameFormatted = schemeNameUnformatted.replace(" ","");
+    	QPixmap preview(getSkinPath(skinName) + "/skin_preview_" + schemeNameFormatted + ".png");
+    	qDebug() << "Looking for: " << getSkinPath(skinName) << "/skin_preview_" << schemeNameFormatted << ".png";
+
```
unfortunately I'm poking in the dark here as call hierarchy doesn't work in my eclipse install

daschuer commented:
I think you need to call  
skinPreviewLabel->setPixmap(m_pSkinLoader->getSkinPreview(newSkin, m_colorScheme));
inside 
slotSetScheme()
as well.

ronso0 commented on src/skin/skinloader.cpp:
```diff
@@ -56,12 +56,25 @@ QString SkinLoader::getSkinPath(const QString& skinName) const {
+    	qDebug() << "Looking for: "<<getSkinPath(skinName)<<"/skin_preview_"<<schemeNameFormatted<<".png";
+
+    } else {
+//    	QPixmap preview(getSkinPath(skinName) + "/skin_preview.png");
```
`Debug [Main]: Looking for: "[install path]/share/mixxx/skins/Deere" /skin_preview.png`
file is there but it still shows the preview placeholder :\

ronso0 committed:
fix pixmap loading

ronso0 committed:
add fake preview images

ronso0 commented:
Yeeehaaa! It's working.

I'll remove the leftovers as soon as I've tested this.

ronso0 commented on src/skin/skinloader.cpp:
```diff
@@ -56,12 +56,19 @@ QString SkinLoader::getSkinPath(const QString& skinName) const {
-QPixmap SkinLoader::getSkinPreview(const QString& skinName) const {
-    QPixmap preview(getSkinPath(skinName) + "/preferences_preview_screenshot.png");
+QPixmap SkinLoader::getSkinPreview(const QString& skinName, const QString& schemeName) const {
+	QPixmap preview(":/images/skin_preview_placeholder.png");
```
I wonder if it's better to initialize `preview` empty and load the default image only when necessary, instead of loading the default and then overwriting it. Probably doesn't make a difference performance-wise but it seems smarter.
?

ronso0 commented on src/preferences/dialog/dlgprefinterface.cpp:
```diff
@@ -191,14 +197,17 @@ void DlgPrefInterface::slotUpdateSchemes() {
-            if (schlist[i] == selectedScheme) {
+            if (schlist[i] == configScheme) {
                 ComboBoxSchemeconf->setCurrentIndex(i);
+                m_colorScheme = configScheme;
```
The only case where the preview is empty after start is when the skin and scheme from mixxx.cfg don't match.
The dropdown and GUi load the first scheme from skin.xml, though.
Is it worth catching that here with another if switch?

ronso0 committed:
initialize empty QPixmap, then load preview images...

Also, when DlgPrefInterface is set up call slotUpdateSchemes after
configured skin is picked to have preview pictures loaded when
Preferences > Interface is opened for the first time after start.

daschuer commented:
> I wonder if it's better to initialize preview empty and load the default image only when necessary, instead of loading the default and then overwriting it. Probably doesn't make a difference performance-wise but it seems smarter, like it was done before.

If you ask, go for the smarter version. In general understandable code is the prio in such performance uncritical places.

daschuer commented:
> Is it worth catching that here with another if switch to show the correct preview?

Yes, this is worth. I think if the compbobox I'd responsible for setting the default skin it is kind of odd though. Can we decide for a default colour scheme before? (If that is worth the effort? I need to look into the code)

daschuer commented:
> If that is worth the effort? I need to look into the code.

Yes, the code should fall back to the first schema if no matching is found. However it should not alter the preferences because it might be a new color schema an the user might test a old version of Mixxx.
A comment in the loop about this issue is IMHO sufficient, no need for an extra default calculation loop.

ronso0 commented on src/skin/skinloader.cpp:
```diff
@@ -56,12 +56,19 @@ QString SkinLoader::getSkinPath(const QString& skinName) const {
+    if (!schemeName.isEmpty()) {
+    	QString schemeNameUnformatted = schemeName;
+    	QString schemeNameFormatted = schemeNameUnformatted.replace(" ","");
+	preview = (getSkinPath(skinName) + "/skin_preview_" + schemeNameFormatted + ".png");
```
is there a smarter/quicker way to get rid of whitespaces in `schemeName`?

daschuer commented:
> As long as the user doesn't hit Apply/Okay after tinkering with the drop-down boxes nothing is changed, right?


Yes, correct.

ronso0 committed:
add comments, fix indentation

ronso0 committed:
add real preview images

nopeppermint commented:
nice !

- the preview pictures are not perfectly align with the drop down menus, in my opinion they should go a little bit more to the right
![settings](https://user-images.githubusercontent.com/3403218/51935501-5bcbd700-2406-11e9-85f8-657b1511f5c2.png)

- the preview for skins without color scheme is not working, i just see the empty preview there (latenigh/deere/deere64sampler)

- the Tango64Sampler and Tango ClubTwist preview is not the same, I thought the difference between Tango and Tango64Sampler is only the amount of samplers available..

ronso0 commented:
> the preview pictures are not perfectly align with the drop down menus, in my opinion they should go a little bit more to the right

That's always been the case, but i can try to fix it.

> the preview for skins without color scheme is not working, i just see the empty preview there

whoopsie! I introduced a regression when cleaning up. Will fix.

> the Tango64Sampler and Tango ClubTwist preview is not the same, I thought the difference between Tango and Tango64Sampler is only the amount of samplers available..

good catch, stupid mistake during the finish.. will add it.

ronso0 committed:
fix preview for skins without color schemes

ronso0 commented:
> the preview pictures are not perfectly align with the drop down menus, in my opinion they should go a little bit more to the right

When the Preferences window is at minimal size (no orizontal scrollbar) the preview stands out equally at the left & right side. As soon as the window is enlarged the comboboxes grow but not the preview (which is fine). I guess I won't fix that unless I find a quick solution.

ronso0 committed:
unify preview images

ronso0 committed:
align left edge of preview image with comboboxes

nopeppermint commented:
looks good now !

daschuer commented:
Works good. Do you still plan to clean up the commits? 
You may consider to merge https://github.com/mixxxdj/mixxx/pull/2005 in this case.

daschuer commented:
LGTM

ronso0 commented:
yes, I want to clean that mess next week. 
are we in a hurry for 2.3?

daschuer commented:
no ... we can wait.

ronso0 commented:
commits cleaned up, merged #2005

daschuer commented:
LGTM Thank you very much. 
I will fix the mall Codacy issue post merge.

ronso0 commented:
niiiice thx!

ronso0 commented:
before the next release I'll update the screenshots, i.e. to reflect the actual fx GUI
